src/plugins/cloudMetadata.nim

```diff
@@ -0,0 +1,335 @@
+##
+## Copyright (c) 2023, Crash Override, Inc.
+##
+## This file is part of Chalk
+## (see https://crashoverride.com/docs/chalk)
+##
+
+## Query common AWS metadata va IMDSv2
+
+import httpclient, net, uri
+import std/strutils, std/json
+import nimutils/sinks
+import ../config, ../plugin_api, ../chalkjson, ./procfs
+
+const
+  awsBaseUri     = "http://169.254.169.254/latest/"
+  awsMdUri       = awsBaseUri & "meta-data/"
+  awsDynUri      = awsBaseUri & "dynamic/"
+  # special keys for special processing
+  AWS_IDENTITY_CREDENTIALS_SECURITY_CREDS = "_AWS_IDENTITY_CREDENTIALS_EC2_SECURITY_CREDENTIALS_EC2_INSTANCE"
+
+proc getAwsToken(): Option[string] =
+  let
+    uri      = parseURI(awsBaseUri & "api/token")
+    hdrs     = newHttpHeaders([("X-aws-ec2-metadata-token-ttl-seconds", "10")])
+    client   = newHttpClient(timeout = 250) # 1/4 of a second
+    response = client.safeRequest(url = uri, httpMethod = HttpPut, headers = hdrs)
+
+  if response.status[0] != '2':
+    trace("Could not retrieve IMDSv2 token from: " & $uri)
+    return none(string)
+
+  trace("Retrieved AWS metadata token")
+  return some(response.bodyStream.readAll().strip())
+
+proc hitAwsEndpoint(path: string, token: string): Option[string] =
+  let
+    uri      = parseUri(path)
+    hdrs     = newHttpHeaders([("X-aws-ec2-metadata-token", token)])
+    client   = newHttpClient(timeout = 250) # 1/4 of a second
+    response = client.safeRequest(url = uri, httpMethod = HttpGet, headers = hdrs)
+
+  if response.status[0] != '2':
+    trace("With valid IMDSv2 token, could not retrieve metadata from: " & $uri)
+    return none(string)
+
+  trace("Retrieved AWS metadata " & uri.path)
+  result = some(response.bodyStream.readAll().strip())
+  if not result.isSome():
+    # log failing keys in trace mode only as some are expected to be absent
+    trace("With valid IMDSv2 token, could not retrieve metadata from: " & uri.path)
+
+proc hitAzureEndpoint(path: string): Option[string] =
+  let
+    uri      = parseUri(path)
+    hdrs     = newHttpHeaders([("Metadata", "true")])
+    client   = newHttpClient(timeout = 250) # 1/4 of a second
+    response = client.safeRequest(url = uri, httpMethod = HttpGet, headers = hdrs)
+
+  if response.status[0] != '2':
+    trace("Could not retrieve Azure metadata from: " & $uri)
+    return none(string)
+
+  trace("Retrieved Azure metadata " & uri.path)
+  result = some(response.bodyStream.readAll().strip())
+  if not result.isSome():
+    # log failing keys in trace mode only as some are expected to be absent
+    trace("Got empty Azure metadata from: " & uri.path)
+
+proc hitGcpEndpoint(path: string): Option[string] =
```

Comment:
If these are all three the same except for the error message (and perhaps the IP) then should parameterize it instead of keeping 3 copies of the code.